mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs

```diff
@@ -0,0 +1,872 @@
+                // Signal that the connection is up and ready.
+                connection_ready_signal.send(was_already_connected).ok();
+                if was_already_connected {
+                    // Gracefully return if some other thread is already connected using `connection`.
+                    // Up until this point, we didn't alter `connection` in any way.
+                    // Note that the other active connection is the one which will reply to the version querying
+                    // and we will still exit `establish_connection_loop` without any errors.
+                    return;
+                }
+
+                event_handlers.on_connected(&address).ok();
+                info!("{address} is now connected");
+
+                let err = select! {
+                    e = no_connection_timeout_f => e,
+                    e = recv_f => e,
+                    e = send_f => e,
+                };
+
+                error!("{address} connection dropped due to: {err:?}");
+                event_handlers.on_disconnected(&address).ok();
+                // FIXME: This will call abort all spawned tasks. INCLUDING THIS ONE WE ARE RUNNING OFF OF.
+                connection.disconnect(Some(err)).await;
+            }
+        };
+        // Start the connection loop on a weak spawner.
+        connection.weak_spawner().spawn(connection_loop);
+
+        // Wait for the connection to be ready before querying the version.
+        let now = Instant::now();
+        match wait_for_connection_ready.timeout_secs(timeout).await {
+            // The connection is already connected. Don't query for the version because Electrum doesn't
+            // like when a client queries for the version multiple times.
+            Ok(Ok(true)) => return Ok(()),
+            // We just connected, need to query for the server version, continue.
+            Ok(Ok(false)) => {},
+            Ok(Err(_)) => {
+                disconnect_and_return!(
+                    Temporary,
+                    format!("Connection ready signal was dropped, weak spawner failed to spawn the connection loop"),
+                    connection,
+                    event_handlers
+                );
+            },
+            Err(_) => {
+                disconnect_and_return!(
+                    Temporary,
+                    format!("Timed out ({timeout}s) while waiting for the connection to be ready"),
+                    connection,
+                    event_handlers
+                );
+            },
+        }
+        // This is the remaining timeout for version checking.
+        let timeout = (timeout - now.elapsed().as_secs_f64()).max(0.0);
+
+        // FIXME: Looks like electrum will return "unsupported protocol version" if we give it some unsupported range.
+        // When then do we need to have version negotiation logic? Electrum won't ever give out of range protocol version, right?
```

Comment:
I am not sure I completely get your question, we support versions 1.2 to 1.4 which we give here, if the electrum we try to connect to uses 1.1 version for instance it will return an error https://electrumx.readthedocs.io/en/latest/protocol-methods.html#server-version

Reply:
Yup, my problem is with `negotiate_version()`. if `negotiate_version()` is true this means we will try to check that the server's chosen version falls in out acceptable range (`client.protocol_version().contains(&version_f32)`). but the server will never send a version outside this range in the first place, it will either send a version in range or error with "unsupported protocol version".